velox/functions/prestosql/types/JsonType.cpp

```diff
@@ -509,292 +511,413 @@ void castToJsonFromRow(
   });
 }
 
-// Write object to writer at the current offset.
-template <TypeKind kind>
-FOLLY_ALWAYS_INLINE void castFromJsonTyped(
-    const folly::dynamic& /*object*/,
-    exec::GenericWriter&
-    /*writer*/) {
-  VELOX_NYI(
-      "Casting from JSON to {} is not supported.", TypeTraits<kind>::name);
+template <typename T>
+simdjson::simdjson_result<T> fromString(const std::string_view& s) {
+  auto ans = folly::tryTo<T>(s);
+  if (ans.hasError()) {
+    return simdjson::INCORRECT_TYPE;
+  }
+  return std::move(*ans);
 }
 
-// Forward declarations.
-template <>
-FOLLY_ALWAYS_INLINE void castFromJsonTyped<TypeKind::ARRAY>(
-    const folly::dynamic& object,
-    exec::GenericWriter& writer);
-
-template <>
-FOLLY_ALWAYS_INLINE void castFromJsonTyped<TypeKind::MAP>(
-    const folly::dynamic& object,
-    exec::GenericWriter& writer);
-
-template <>
-FOLLY_ALWAYS_INLINE void castFromJsonTyped<TypeKind::ROW>(
-    const folly::dynamic& object,
-    exec::GenericWriter& writer);
-
-template <>
-FOLLY_ALWAYS_INLINE void castFromJsonTyped<TypeKind::VARCHAR>(
-    const folly::dynamic& object,
-    exec::GenericWriter& writer) {
-  if (isJsonType(writer.type())) {
-    // Sort keys to match Presto's behavior.
-    folly::json::serialization_opts opts;
-    opts.sort_keys = true;
-    writer.castTo<Varchar>().append(folly::json::serialize(object, opts));
-  } else if (object.isBool()) {
-    writer.castTo<Varchar>().append(object.asBool() ? "true" : "false");
-  } else {
-    writer.castTo<Varchar>().append(object.asString());
+template <typename To, typename From>
```

Comment:
Would you add comments for this function and explain what are valid To and From types?